packages/contracts-ops/contracts/scripts/UpgradeCallBatches.s.sol

```diff
@@ -0,0 +1,129 @@
+pragma abicoder v2;
+
+/*//////////////////////////////////////////////////////////////
+                                 IMPORTS
+    //////////////////////////////////////////////////////////////*/
+
+// Utilities
+import {Script, console2} from "forge-std/Script.sol";
+
+// Ops libs
+import {Config} from "../Config.sol";
+import {CallBatch} from "../CallBatch.sol";
+
+// Contract for selector
+import {UpgradeBeaconController} from "@nomad-xyz/contracts-core/contracts/upgrade/UpgradeBeaconController.sol";
+
+contract UpgradeCallBatches is Script, Config, CallBatch {
+    /*//////////////////////////////////////////////////////////////
+                            GOVERNANCE CALLS
+    //////////////////////////////////////////////////////////////*/
+
+    bytes upgradeHome;
+    bytes upgradeReplica;
+    bytes upgradeGovRouter;
+    bytes upgradeBridgeRouter;
+    bytes upgradeTokenRegistry;
+    bytes upgradeBridgeToken;
+
+    string configFile;
+    string[] domainNames;
+
+    function printCallBatches(
+        string memory _configFile,
+        string[] memory _domainNames,
+        string memory _localDomainName,
+        bool recovery
+    ) external {
+        localDomainName = _localDomainName;
+        domainNames = _domainNames;
+        configFile = _configFile;
+        setUp();
+        for (uint256 i; i < domainNames.length; i++) {
+            string memory domain = domainNames[i];
+            generateGovernanceCalls(domain);
+        }
+        writeCallBatch(recovery);
+    }
+
+    function setUp() internal {
+        __Config_initialize(configFile);
+        string memory outputFile = "upgradeActions.json";
+        __CallBatch_initialize(
+            localDomainName,
+            getDomainNumber(localDomainName),
+            outputFile,
+            true
+        );
+        console2.log("Governance Actions have been output to", outputFile);
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                       GOVERNANCE CALL GENERATORS
+    //////////////////////////////////////////////////////////////*/
+
+    function generateGovernanceCalls(string memory domain) internal {
+        upgradeHome = abi.encodeWithSelector(
+            UpgradeBeaconController.upgrade.selector,
+            address(homeUpgrade(domain).beacon),
+            address(homeUpgrade(domain).implementation)
+        );
+        string[] memory networks = getNetworks();
+        address replicaImpl;
+        address replicaBeacon;
+        for (uint256 i; i < networks.length; i++) {
+            if (getDomainNumber(networks[i]) != getDomainNumber(domain)) {
+                replicaImpl = address(
+                    replicaOfUpgrade(domain, networks[i]).implementation
+                );
+                replicaBeacon = address(
+                    replicaOfUpgrade(domain, networks[i]).beacon
+                );
+            }
+        }
+        upgradeReplica = abi.encodeWithSelector(
```

Comment:
suggested refactor: 

```
function pushUpgrade(Upgrade _upgrade, string memory domain) private {
        bytes call = abi.encodeWithSelector(
            UpgradeBeaconController.upgrade.selector,
            address(_upgrade.beacon),
            address(_upgrade.implementation)
        );
        uint32 domainNumber = getDomainNumber(domain);
        address beaconController = address(getUpgradeBeaconController(domain));
        if (domainNumber == getDomainNumber(localDomainName)) {
             pushLocal(beaconController, call);
        } else {
             pushRemote(beaconController, call, domainNumber);
        }       
}
```

Reply:
```
pushUpgrade(homeUpgrade(domain), domain);
pushUpgrade(replicaOfUpgrade(domain, getConnections(domain)[0]), domain);
```
... etc

Reply:
note this would be a pretty valid place to store domain to storage, to avoid it being passed as a parameter redundantly all the time. just need to make very sure you are actually updating that storage variable for each loop.

Reply:
This is a very good suggestion. Thanks